icepyx/core/read.py

```diff
@@ -0,0 +1,614 @@
+        # maybe give user some options here about merging parameters?
+        # add a check that wanted variables exists, and create them with defaults if possible (and let the user know)
+        # write tests for the functions!
+
+        # Notes: intake wants an entire group, not an individual variable (which makes sense if we're using its smarts to set up lat, lon, etc)
+        # so to get a combined dataset, we need to keep track of spots under the hood, open each group, and then combine them into one xarray where the spots are IDed somehow (or only the strong ones are returned)
+        # this means we need to get/track from each dataset we open some of the metadata, which we include as mandatory variables when constructing the wanted list
+
+        try:
+            groups_list = list_of_dict_vals(self._read_vars.wanted)
+        except AttributeError:
+            pass
+
+        all_dss = []
+        # DevNote: I'd originally hoped to rely on intake-xarray in order to not have to iterate through the files myself,
+        # by providing a generalized url/source in building the catalog.
+        # However, this led to errors when I tried to combine two identical datasets because the single dimension was equal.
+        # In these situations, xarray recommends manually controlling the merge/concat process yourself.
+        # While unlikely to be a broad issue, I've heard of multiple matching timestamps causing issues for combining multiple IS2 datasets.
+        for file in self._filelist:
+            all_dss.append(
+                self._build_single_file_dataset(file, groups_list)
+            )  # wanted_groups, vgrp.keys()))
+
+        if len(all_dss) == 1:
+            return all_dss[0]
+        else:
+            try:
+                merged_dss = xr.combine_by_coords(all_dss, data_vars="minimal")
+                return merged_dss
+            except ValueError as ve:
+                warnings.warn(
+                    "Your inputs could not be automatically merged using "
+                    f"xarray.combine_by_coords due to the following error: {ve}\n"
+                    "icepyx will return a list of Xarray DataSets (one per granule) "
+                    "which you can combine together manually instead",
+                    stacklevel=2,
+                )
+                return all_dss
+
+    def _build_dataset_template(self, file):
+        """
+        Create the Xarray dataset object templated for the data to be read in.
+
+        It may be possible to expand this function to provide multiple templates.
+        """
+        # NOTE: use the hdf5 library to grab the attr for the product specifier
+        # can ultimately then use it to check against user specified one or merge strategies (or to return a list of ds)
+
+        is2ds = xr.Dataset(
+            coords=dict(
+                gran_idx=["999999"],
+                source_file=(["gran_idx"], [file]),
```

Comment:
Given that the file names are long strings, turn this into an attribute of each "gran_idx" DataArray rather than a coordinate.